### PR TITLE
Crimson/osd: Disable concurrent MOSDMap handling

### PR DESCRIPTION
**Pull request by Matan-B**
Fixes: https://tracker.ceph.com/issues/59165

Note: This PR is originated from #51425 and #51860 attempts to avoid introducing new `seastar::shared_mutex` member. However, in order to avoid unnecessary complexity in the code for non-significant impact on performance the current change was favoured.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/osd/shard_services.cc [hunk 1]
```diff
@@ -421,19 +426,23 @@ seastar::future<> OSDSingletonState::store_maps(ceph::os::Transaction& t,
...
 	o->decode(p->second);
-	logger().info("store_maps osdmap.{}", e);
+	logger().info("store_maps storing osdmap.{}", e);
 	store_map_bl(t, e, std::move(std::move(p->second)));
 	osdmaps.insert(e, std::move(o));
 	return seastar::now();
       } else if (auto p = m->incremental_maps.find(e);
 		 p != m->incremental_maps.end()) {
+	ceph_assert (std::cmp_greater(e, 0u));
+	logger().info("store_maps found osdmap.{} incremental map, "
```
Comment (rzarzynski) [inline src/crimson/osd/shard_services.cc:436]: Does it make sense to log before assertion?

**Review by rzarzynski — APPROVED** [review]

File: src/crimson/osd/shard_services.cc (hunk 1 above)
Comment (Matan-B) [reply src/crimson/osd/shard_services.cc:436]: Yes, updated.

File: src/crimson/osd/osd.cc [hunk 2]
```diff
@@ -855,6 +855,16 @@ bool OSD::require_mon_peer(crimson::net::Connection *conn, Ref<Message> m)
 
 seastar::future<> OSD::handle_osd_map(crimson::net::ConnectionRef conn,
                                       Ref<MOSDMap> m)
+{
```
Comment (athanatos) [inline src/crimson/osd/osd.cc:857]: Let's leave a comment explaining the rationale:

```
  /* Ensure that only one MOSDMap is processed at a time.  Allowing concurrent
   * processing may eventually be worthwhile, but such an implementation would
   * need to ensure (among other things)
   * 1. any particular map is only processed once
   * 2. PGAdvanceMap operations are processed in order for each PG
   * As map handling is not presently a bottleneck, we stick to this simpler invariant
   * for now.
   * See https://tracker.ceph.com/issues/59165
   */
  return handle_osd_map_lock.lock().then([=, this] {
    return _handle_osd_map(conn, m);
  }).then([=, this] {
    return handle_osd_map_lock.unlock();
  });
```

Commit message should be:

```
crimson/osd/osd: serialize MOSDMap message handling

See included comment for rationale.

Fixes: https://tracker.ceph.com/issues/59165
Signed-off-by: Matan Breizman <mbreizma@redhat.com>
```

A future reader will want to know *why* there's a mutex there, especially if it one day map processing becomes a bottleneck.

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/crimson/osd/osdmap_gate.cc [hunk 3]
```diff
@@ -54,6 +54,9 @@ seastar::future<epoch_t> OSDMapGate<OSDMapGateTypeV>::wait_for_map(
 
 template <OSDMapGateType OSDMapGateTypeV>
 void OSDMapGate<OSDMapGateTypeV>::got_map(epoch_t epoch) {
+  if (current >= epoch) {
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (Matan-B) [reply src/crimson/osd/osdmap_gate.cc:59]: `epoch` and `current` are both 0 when the osd is starting.  Beside that case they shouldn't be, I added an assert instead.

File: src/crimson/osd/osd.cc (hunk 2 above)
Comment (Matan-B) [reply src/crimson/osd/osd.cc:857]: Updated, thank you.

**Review by athanatos — APPROVED** [review]

File: src/crimson/osd/osd_operations/pg_advance_map.cc [hunk 4]
```diff
@@ -62,7 +63,7 @@ seastar::future<> PGAdvanceMap::start()
   return enter_stage<>(
     pg->peering_request_pg_pipeline.process
   ).then([this] {
-    from = pg->get_osdmap_epoch();
+    ceph_assert(from == pg->get_osdmap_epoch());
```
Comment (Matan-B) [inline src/crimson/osd/osd_operations/pg_advance_map.cc:66]: This assertion was true on several tests, marked as DNM until resolved.

**Comment by Matan-B** [comment]
Tests look good.
https://pulpito.ceph.com/matan-2023-06-19_11:10:19-crimson-rados-wip-matanb-crimson-single-mosdmap-handling-v3-distro-crimson-smithi/

File: src/crimson/osd/osdmap_gate.cc [hunk 5]
```diff
@@ -54,6 +54,10 @@ seastar::future<epoch_t> OSDMapGate<OSDMapGateTypeV>::wait_for_map(
 
 template <OSDMapGateType OSDMapGateTypeV>
 void OSDMapGate<OSDMapGateTypeV>::got_map(epoch_t epoch) {
+  if (!epoch) {
```
Comment (athanatos) [inline src/crimson/osd/osdmap_gate.cc:57]: epoch == 0

File: src/crimson/osd/osd_operations/pg_advance_map.cc [hunk 6]
```diff
@@ -65,6 +65,8 @@ seastar::future<> PGAdvanceMap::start()
   return enter_stage<>(
     pg->peering_request_pg_pipeline.process
   ).then([this] {
+    // `from` epoch may vary between PGs,
+    // see: https://tracker.ceph.com/issues/61744
```
Comment (athanatos) [inline src/crimson/osd/osd_operations/pg_advance_map.cc:69]: Include the explanation here directly.

**Review by athanatos — CHANGES_REQUESTED** [review]

**Review by athanatos — APPROVED** [review]

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2023-06-21_12:00:29-crimson-rados-wip-matanb-crimson-single-mosdmap-handling-distro-crimson-smithi/

7309664, 7309667: https://tracker.ceph.com/issues/61651


[7309664](https://pulpito.ceph.com/matan-2023-06-21_12:00:29-crimson-rados-wip-matanb-crimson-single-mosdmap-handling-distro-crimson-smithi/7309664)

**Pull request merged**